docs/dev/writing-code.md

```diff
@@ -192,18 +201,195 @@ TODO: document when to use future vs callback
 
 # Lifetime Management
 
-TODO: Write me.
-Talk about Garbage Collector vs Reference Counting. Have pictures.
-Talk about capsules. Talk about NativeResource. Warn about cycles.
-Talk about how our tests can and cannot check for leaks.
-Talk about which classes require a `close()` function, and which don't.
+## Terminology
+
+### Strong References / Reference Counting
+
+A "strong reference" is one that keeps an object alive by incrementing its reference count.
+When all references to an object are released (reference count goes to zero), it gets cleaned up.
+
+In Python code, every variable is a strong reference to an object.
+When the variable goes away, the reference is released.
+C code can create a strong reference to a Python object by calling `Py_INCREF(x)`,
+and release the reference by calling `Py_DECREF(x)`.
+Note that EVERYTHING in Python is an object: even functions, even numbers, even `None`.
+
+In the aws-c libraries, you create a strong reference to a struct by calling
+its `_acquire(x)` function, and release by calling its `_release(x)` function.
+Within the aws-c libraries, structs keep each other alive as long as necessary using these functions.
+Not every struct in the aws-c libraries has `_acquire(x)` and `_release(x)` functions,
+only heap-allocated structs with complex or unpredictable lifetimes.
+Every struct bound to a Python class is considered to have an unpredictable lifetime.
+
+### Reference Cycle
+
+A "reference cycle" is when a circle of strong references is created.
+Reference cycles cause memory to leak because the reference counts never get to zero.
+
+Python has a [garbage collector](https://devguide.python.org/internals/garbage-collector)
+that can detect and clean up reference cycles among normal Python objects.
+HOWEVER, any cycle involving a `Py_INCREF(x)` from C creates an undetectable cycle.
+You MUST NOT create reference cycles when designing bindings.
+
+### Capsule
+
+[PyCapsule](https://docs.python.org/3/extending/extending.html#using-capsules)
+lets us bind the lifetime of a C struct to the lifetime of a Python object.
+It's a Python object that holds a C pointer and a "destructor" function pointer.
+When Python cleans up the `PyCapsule`, the destructor function will be called.
+
+## Bindings Design
+
+Let's look at the bindings for `aws_event_loop_group` (our I/O thread pool).
+This diagram shows the strong references between objects in Python and C:
+
+![Diagram of Simple Binding](./binding-simple.svg)
+
+Description of parts:
+*   `aws_event_loop_group` - The underlying native implementation struct,
+    which knows nothing about Python.
+    *   Lives in C library: `aws-c-io`
+    *   Header file: `crt/aws-c-io/include/aws/io/event_loop.h`
+*   `event_loop_group_binding` - The "bindings" struct.
+    Holds a strong reference to the underlying native implementation (usually in a member variable named "native")
+    *   Lives in Python/C extension module: `_awscrt`
```

Comment:
This is probably because when calling things in the crt, you use the '_awscrt' module.
e.g. to call subscribe from python I use '_awscrt.mqtt5_client_subscribe()' and not 'aws-crt-python.mqtt5_client_subscribe()'